schemachanger: inject all dependencies

postamar opened the pull request:
Although the declarative schema changer had been written with
    dependency injection in mind, the level of abstraction was still too
    low for straightforward end-to-end testing. The schema changer doesn't
    care about kv.Txn, jobs.JobRegistry, etc. Fundamentally, here's how it
    interacts with the rest of the system:
    1. The builder in scbuild:
      a. The builder queries the catalog for the current descriptor and
         namespace state.
      b. The builder also needs a valid tree.SemaContext and
         tree.EvalContex, but only in a limited way. DDLs don't require the
         evaluation of built-in functions or the like, meaning that suitable
         context objects can be built with few indirect dependencies like
         the codec, session data, etc.
    2. The executor in scexec:
      a. Naturally the executor also needs access to the catalog for reading
         and writing. In the scmutationexec package these dependencies had
         already been abstract enough, meaning there was little left to do
         except formalize read-only vs. mutating dependencies a bit.
      b. The same holds for the backfill operations, with the
         IndexBackfiller and JobProgressTracker interfaces.
      c. The executor needs to know about testing knobs and other debug
         info.
    
    In addition to making dependencies explicit as described above in
    scbuild.Dependencies and scexec.Dependencies interfaces, this commit
    also makes dependencies explicit for the actual running of the schema
    changer, to allow end-to-end testing. Concretely, we added a layer of
    abstraction around the execution of each phase:
    1. For scop.StatementPhase, the schema changer receives a target
       state, runs the planner to generate ops-stages, and executes these
       ops.
    2. For scop.PreCommitPhase, it does exactly that, but also creates and
       enqueues a job, which will deal with the subsequent
       scop.PostCommitPhase asynchronously. This job creation depends on
       things like the job registry, which we don't care for if we're doing
       end-to-end testing: in that case we'll want to run the post-commit
       phase synchronously after this one.
    3. For scop.PostCommitPhase, it's like the previous phases, except in
       the context of the Resume method of the schema change job, which
       requires a bit of plumbing to update the progress of the job itself,
       and also to update the descriptors affected by the job, because those
       hold a reference to this job which serves as a lock, blocking other
       concurrent schema changes.
    
    This new layer of abstraction is defined in the new scrun package.
    The actual implementations of the dependency interfaces are all moved to
    a new scdeps package.
    
    This commit also introduces test implementations of these dependency
    interfaces, which so far are used in the builder tests alongside
    the regular implementations. Adoption in the executor tests, as well as
    end-to-end testing of the declarative schema changer, will be added in
    a subsequent commit.
    
    Release note: None

Commit:
schemachanger: inject all dependencies

Although the declarative schema changer had been written with
dependency injection in mind, the level of abstraction was still too
low for straightforward end-to-end testing. The schema changer doesn't
care about kv.Txn, jobs.JobRegistry, etc. Fundamentally, here's how it
interacts with the rest of the system:
1. The builder in scbuild:
  a. The builder queries the catalog for the current descriptor and
     namespace state.
  b. The builder also needs a valid tree.SemaContext and
     tree.EvalContex, but only in a limited way. DDLs don't require the
     evaluation of built-in functions or the like, meaning that suitable
     context objects can be built with few indirect dependencies like
     the codec, session data, etc.
2. The executor in scexec:
  a. Naturally the executor also needs access to the catalog for reading
     and writing. In the scmutationexec package these dependencies had
     already been abstract enough, meaning there was little left to do
     except formalize read-only vs. mutating dependencies a bit.
  b. The same holds for the backfill operations, with the
     IndexBackfiller and JobProgressTracker interfaces.
  c. The executor needs to know about testing knobs and other debug
     info.

In addition to making dependencies explicit as described above in
scbuild.Dependencies and scexec.Dependencies interfaces, this commit
also makes dependencies explicit for the actual running of the schema
changer, to allow end-to-end testing. Concretely, we added a layer of
abstraction around the execution of each phase:
1. For scop.StatementPhase, the schema changer receives a target
   state, runs the planner to generate ops-stages, and executes these
   ops.
2. For scop.PreCommitPhase, it does exactly that, but also creates and
   enqueues a job, which will deal with the subsequent
   scop.PostCommitPhase asynchronously. This job creation depends on
   things like the job registry, which we don't care for if we're doing
   end-to-end testing: in that case we'll want to run the post-commit
   phase synchronously after this one.
3. For scop.PostCommitPhase, it's like the previous phases, except in
   the context of the Resume method of the schema change job, which
   requires a bit of plumbing to update the progress of the job itself,
   and also to update the descriptors affected by the job, because those
   hold a reference to this job which serves as a lock, blocking other
   concurrent schema changes.

This new layer of abstraction is defined in the new scrun package.
The actual implementations of the dependency interfaces are all moved to
a new scdeps package.

This commit also introduces test implementations of these dependency
interfaces, which so far are used in the builder tests alongside
the regular implementations. Adoption in the executor tests, as well as
end-to-end testing of the declarative schema changer, will be added in
a subsequent commit.

Release note: None

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

I'm pleased with this. I like the strategy of pulling out test state from the cluster in hopes of one day cutting the cluster out of the picture. Only nits from me.

> Reviewed 24 of 43 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71898#-:-Mn1Ok0AE50bOxYAZdV7:blqsd2i)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @fqazi and @postamar)

___
*[pkg/sql/schemachanger/scbuild/dependencies.go, line 83 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71898#-Mn16HvKCi1W23p8Ed8l:-Mn16HvKCi1W23p8Ed8m:b-8emz88) ([raw file](https://github.com/cockroachdb/cockroach/blob/29744661e6522e592b3bee051e01dedc0327fa12/pkg/sql/schemachanger/scbuild/dependencies.go#L83)):*
> ```Go
> 		ctx context.Context, descriptor catalog.Descriptor, privilege privilege.Kind,
> 	) error
> 	// HasAdminRole verifies if a user has an admin role
> ```

I know you said don't, but I have to.

nit: period
___
*[pkg/sql/schemachanger/scdeps/run_deps.go, line 89 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71898#-Mn14RAX3OKaK4o-5x5d:-Mn14RAX3OKaK4o-5x5e:bqzjsb6) ([raw file](https://github.com/cockroachdb/cockroach/blob/29744661e6522e592b3bee051e01dedc0327fa12/pkg/sql/schemachanger/scdeps/run_deps.go#L89)):*
> ```Go
> 		jobRegistry:       jobRegistry,
> 		job:               job,
> ```

uber-nit: why this space?
___
*[pkg/sql/schemachanger/scdeps/sctestdeps/test_state.go, line 40 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71898#-Mn1OKE-3_NQ5xiSuGxq:-Mn1OKE-3_NQ5xiSuGxr:b-ysj84z) ([raw file](https://github.com/cockroachdb/cockroach/blob/29744661e6522e592b3bee051e01dedc0327fa12/pkg/sql/schemachanger/scdeps/sctestdeps/test_state.go#L40)):*
> ```Go
> 	currentDatabase string
> 	phase           scop.Phase
> 	sessionData     sessiondata.SessionData
> ```

this too can now be serialized from sql, see `crdb_internal.serialize_session`
___
*[pkg/sql/schemachanger/scexec/dependencies.go, line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71898#-Mn14ea-EB7h8M15FfOC:-Mn14ea-EB7h8M15FfOD:b-59ugrh) ([raw file](https://github.com/cockroachdb/cockroach/blob/29744661e6522e592b3bee051e01dedc0327fa12/pkg/sql/schemachanger/scexec/dependencies.go#L54)):*
> ```Go
> 	MustReadMutableDescriptor(ctx context.Context, id descpb.ID) (catalog.MutableDescriptor, error)
> 
> 	// NewCatalogChangeBatcher is qquivalent to creating a new kv.Batch for the
> ```

qquivalent


<!-- Sent from Reviewable.io -->

Commit:
descs: honor AvoidSynthetic flag when getting by name

Previously, the AvoidSynthetic object selection flag was only respected
in descriptor lookups by ID. This commit fixes this by extending the
same logic to name lookups.

Presently, there are no lookups by name that use this flag, so the
impact is nil, and this change is purely defensive.

Release note: None

Commit:
catalog: add NewBuilder method to Descriptor

Previously, creating a builder for an arbitrary descriptor object
required importing the catalogkv package, because that's where the
descriptor proto de/serialization mostly happens. This is awkward when
all we want to do is deep-copy a descriptor. This commit addresses this
growing need.

Release note: None

fqazi reviewed (approved):
> Reviewed 43 of 43 files at r1, 18 of 47 files at r2, 43 of 43 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71898)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @adityamaru and @postamar)



<!-- Sent from Reviewable.io -->
